pymedphys/_dicom/anonymise/strategy.py

```diff
@@ -0,0 +1,68 @@
+# Copyright (C) 2020 Stuart Swerdloff, Simon Biggs
+# Copyright (C) 2018 Matthew Jennings, Simon Biggs
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import logging
+
+from pymedphys._imports import pydicom
+
+from pymedphys._dicom.constants import PYMEDPHYS_ROOT_UID
+
+REPLACEMENT_TO_VR_MAP = {
+    "Anonymous": ["AE", "LO", "LT", "PN", "SH", "ST"],
+    "100Y": ["AS"],
+    "ANON": ["CS"],
+    "20190303": ["DA"],
+    "12345678.9": ["DS"],
+    "20190303000900.000000": ["DT"],
+    (0).to_bytes(2, "little"): ["OB", "OB or OW", "OW"],
+    "000900.000000": ["TM"],
+    PYMEDPHYS_ROOT_UID: ["UI"],
+    12345: ["US"],
+}
+
+VR_TO_REPLACEMENT_MAP = {}
+for value, keys in REPLACEMENT_TO_VR_MAP.items():
+    for key in keys:
+        VR_TO_REPLACEMENT_MAP[key] = value
+
+
+def _get_vr_anonymous_hardcode_replacement_value(current_value, value_representation):
+    """A single dispatch function that is used for any VR with the current_value of the element ignored
+    This is the default for the replacement strategy
+    """
+    if current_value is None:
+        # TODO: Decide if we actually want to be doing this. I haven't
```

Comment:
Personally, I don't like replacing an empty value with a hardcoded anonymisation value.  I don't think it makes the data any more anonymous and it risks interoperability issues or worse, DICOM conformance violation.
But one of the nice things about using the strategy (pattern) is substitution of an alternative strategy is straight forward.  So a different strategy is "hardcode but don't replace empty values", and empty could be either "no value" or "empty string".
I tend to avoid changing things unless I know they are wrong, instead I tend to offer an alternative.  One does end up with a lot of cruft after a while, but it avoids accidentally breaking things because someone had an expectation down the line that maybe they should not have.

Reply:
> One does end up with a lot of cruft after a while, but it avoids accidentally breaking things because someone had an expectation down the line that maybe they should not have.

At the moment, given we are version 0.*, we are quite free to break expectations, especially to improve code quality and keep cruft out, if possible. In the future, we are still free to break expectations, it just has to come with deprecation warnings and the like.

I think for PyMedPhys to continue to stand under its own weight into the future, given the size of the development team, we simply cannot afford to let cruft hang around.

I really like the pragmatic programmer book, on page 27 onwards he describes "software entropy" and how some teams manage to successfully combat it.

https://www.cin.ufpe.br/~cavmj/104The%20Pragmatic%20Programmer,%20From%20Journeyman%20To%20Master%20-%20Andrew%20Hunt,%20David%20Thomas%20-%20Addison%20Wesley%20-%201999.pdf#page=27

Excerpt:

> While software development is immune from almost all physical laws, entropy hits us hard.
Entropy is a term from physics that refers to the amount of "disorder" in a system.
Unfortunately, the laws of thermodynamics guarantee that the entropy in the universe
tends toward a maximum. When disorder increases in software, programmers call it
"software rot."
>
> There are many factors that can contribute to software rot. The most important one seems
to be the psychology, or culture, at work on a project. Even if you are a team of one, your
project's psychology can be a very delicate thing. Despite the best laid plans and the best
people, a project can still experience ruin and decay during its lifetime. Yet there are other
projects that, despite enormous difficulties and constant setbacks, successfully fight
nature's tendency toward disorder and manage to come out pretty well.
>
> What makes the difference?


-----

> Personally, I don't like replacing an empty value with a hardcoded anonymisation value. I don't think it makes the data any more anonymous and it risks interoperability issues or worse, DICOM conformance violation.

I agree, it has stepped beyond its task, a function shouldn't go and do more than one expects it to. And it is confusing that they appear when they weren't there before.

I say we remove it and put a note in the changelog to warn people as such.